### ISSUE TITLE
Pipeline for Online Directional Forecasts 

### ISSUE THREAD

**Issue by ViktorKaz** [alan-turing-institute/sktime#763]
# Problem

In financial applications it is common to be interested in predicting at time $t$ the direction of change for an asset at time $t+m$, where $m \geq 1$ is a future time period. Generally speaking, there are three possibilities. In the time period $[t,t+m]$ either the price can go up, down or there will be no change. 

This problem can be decomposed in several smaller, more manageable ones, namely:

1. Dimensionality Reduction and/or Feature Extraction
1. Feature Transformation
1. Changepoint Detection
1. Annotation (Labelling)
1. Temporal cross validation and estimator tuning

 
# Solution
Ideally, we would have a pipeline which chains together the different elements of the algorithm. In contrast to `sklearn` pipelines we will need the capability to chain the different steps in the pipeline not only sequentially but in any giver order. Furthermore, we will need a functionality for transforming and processing both the features and the labels in our dataset.

# Alternatives
`mlfinlab` has functionality with which directional forecasts can be made. However, it is not implemented as a pipeline but rather as a collection of loose functions organised in modules. Furthermore, `mlfinlab` is not open source. 

Other open source toolboxes that tackle at least part of the problem include:
* `ruptures` deals with change point detection, 
* `ADTK` with anomaly detection (specific case of change point detection), 
* `PyOD` was covers outlier detection (specific case of change point detection)

# Additional context

## Background: `mlfinab` example

```Python
#feature reduction and annotation
daily_vol = ml.util.volatility.get_daily_vol(close=dollar_bars['close'], lookback=5)
cusum_events = ml.filters.cusum_filter(dollar_bars['close'], threshold=-0.1)
vertical_barriers = ml.labeling.add_vertical_barrier(t_events=cusum_events, close=dollar_bars['close'], num_days=5)
triple_barrier_events = ml.labeling.get_events(close=dollar_bars['close'],
                                                t_events=cusum_events,
                                                pt_sl=[1, 1], # sets the width of the barriers
                                                target=daily_vol * 2, # twice the daily vol
                                                min_ret=0.005,
                                                num_threads=4,
                                                vertical_barrier_times=vertical_barriers,
                                                side_prediction=None,
                                                verbose=False)
triple_labels = ml.labeling.get_bins(triple_barrier_events, dollar_bars['close'])

#tune and fit an estimator
train_idx, test_idx = train_test_split(np.arange(dataset.shape[0]), shuffle=False, test_size=0.25)
cv_gen = ml.cross_validation.PurgedKFold(samples_info_sets=samples_info_sets['t1'].iloc[train_idx], n_splits=5, pct_embargo=0.01)
gs = GridSearchCV(estimator=estimator, param_grid=param_grid, scoring='neg_log_loss', cv=cv_gen)
train_dts = dataset.iloc[train_idx]
X = train_dts.iloc[:,0:lookback]
y = train_dts['bin']

gs.fit(X,y)
```

## Background: `sktime` transformers

Below is the `sktime` base transformer class
```Python

class BaseTransformer(BaseEstimator):
    """Transformer base class"""

    def __init__(self):
        super(BaseTransformer, self).__init__()

    def fit(self, Z, X=None):
        """
        Fit transformer to X and y.
        By default, fit is empty. Fittable transformations overwrite fit method.
        Parameters
        ----------
        Z : TimeSeries
            Data to be transformed
        X : TimeSeries, optional (default=None)
            Extra data
        Returns
        -------
        self : a fitted instance of the estimator
        """
        self._is_fitted = True
        return self

    def transform(self, Z, X=None):
        """Transform data. Returns a transformed version of X."""
        raise NotImplementedError("abstract method")

    def fit_transform(self, Z, X=None):
        """Fit to data, then transform it.
        Fits transformer to X and y with optional parameters fit_params
        and returns a transformed version of X.
        Parameters
        ----------
        Z : pd.DataFrame, pd.Series or np.ndarray
            Data to be transformed
        X : pd.Series or np.ndarray, optional (default=None)
            Target values of data to be transformed.
        Returns
        -------
        Zt : pd.DataFrame, pd.Series or np.ndarray
            Transformed data.
        """
        # Non-optimized default implementation; override when a better
        # method is possible for a given algorithm.
        if X is None:
            # Fit method of arity 1 (unsupervised transformation)
            return self.fit(Z).transform(Z)
        else:
            # Fit method of arity 2 (supervised transformation)
            return self.fit(Z, X).transform(Z)
```

Private transformer classes inherit from base transformer. The defining characteristic of the transformers are their input and output types.

```Python
class _SeriesToPrimitivesTransformer(BaseTransformer):
    """Transformer base class for series to primitive(s) transforms"""

    def transform(self, Z: Series, X=None) -> Primitives:
        raise NotImplementedError("abstract method")


class _SeriesToSeriesTransformer(BaseTransformer):
    """Transformer base class for series to series transforms"""

    def transform(self, Z: Series, X=None) -> Series:
        raise NotImplementedError("abstract method")


class _PanelToTabularTransformer(BaseTransformer):
    """Transformer base class for panel to tabular transforms"""

    def transform(self, X: Panel, y=None) -> Tabular:
        raise NotImplementedError("abstract method")


class _PanelToPanelTransformer(BaseTransformer):
    """Transformer base class for panel to panel transforms"""

    def transform(self, X: Panel, y=None) -> Panel:
        raise NotImplementedError("abstract method")
```

## Pipeline: proposed user journey

```Python
pipeline = OnlineUnsupervisedPipeline(
    [
        ('feature_extraction', DollarBars(price_col ='Price', volume_col ='Volume'), input_values={'X':'original_series'} ), # input_values={'X':'original_series'} can be omitted
        ('changepoint_detection',CUSUM(), input_values={'X':'feature_extraction'}),
        ('labeling', TripleBarrier(price_col='closing_price'), input_values={ 'change_points':'changepoint_detection', 'prices':'feature_extraction'}), 
        ('build_dataset', DatasetBuilder(lookback=20), input_values={ 'untransformed_series':'original_series','changepoint_detection':'changepoint_detection'}), 
        ('estimator',SVC(), input_values={'X':'build_dataset','y':'labelling'})
    ]
)
pipeline.fit(X)

```

## Scientific Types and Templates

The pipeline is comprised of different steps. We propose that each of the algorithms comprising the pipeline is a `transformer`, except the last one which will be an `estimator`.

The scientific type of the transformers will be defined by reference to the input and output types that the transformer handles. The existing `sktime` templates will be used and new ones will be added.

```Python
class __ComplexToTabularTransformer(BaseTransformer):
	"""
	New transformer abstract base class that takes as inputs two or more different types 
	"""
	def transform (self, inputs) -> Tabular:
		pass
class DollarBars(_PanelToTabularTransformer):
	"""
	The input will need to be to the fit method is a Pandas Dataframe with arbitrry names of the columns. We need to specify the names of the columns that the algorithm will be looking for.
	
	Parameters
	----------
	price_col : string
	volume_col : sting
	threashold : float
		aggregation threashold
	return_col_names : array
		array with names of columns in the transformed pandas dataframe
	"""
	def __init__(self, price_col, volume_col, threashold, return_col_names=['high_price','low_price','open_price','close_price','num_trades','cumul_dollar_value']):
		pass
		
	
	def fit(self, X):
		"""
		Finds the cut off agregation points in the input dataset X
		
		Paramters
		---------
		X : Pandas DataFrame
			columns: index - timestamps, Price, Volume, etc
		"""
		
		pass
		
		
	def transform(self):
		"""
		Returns Aggregated X
		
		Returns
		-------
			Pandas DataFrame, where index - timestamp, columns - high price, low price, open, close, number of trades cumulative dollar value, etc
		"""
	def fit_transfom(self, X):
		"""
		calls the fit and the transofmr methods sequentially
		"""
		pass
	def update(self,X):
		"""
		Performs update only for the new values since the last run of fit_transform and concatenates the results to the DataFrame previously returned by the `transform` method.
		"""
		pass
		
class CUSUM(_SeriesToSeriesTransformer):
	"""
	 CUSUM is q quality control method designed to detect a shift in the mean value of a measured quantity away from a target value.
	Parameters
	----------
	target_col:string
		name of input column
	threashold : float
		defines shift from target value		
	"""
	def __init__(self,target_col, threashold):
		pass
	def fit(self, X):
		"""
		Paramters
		---------
		X : Pandas Dataframe or Pandas Series
			columns index-timestamps, closing price
		"""
		pass
	def transform(self):
		"""
		Returns
		-------
			array : index values from the input dataset that represent change points
		"""
		pass
	def fit_transfom(self, X):
		"""
		calls the fit and the transofmr methods sequentially
		"""
		pass
	def update(self,X):
		"""
		Performs update only for the new values since the last run of fit_transform and concatenates the results to the DataFrame previously returned by the `transform` method.
		"""
		pass
class TrippleBarrier(_ComplexToTabularTransformer):
	def __init__(self, price_col):
		"""
		Parameters
		----------
		price_col : string
			name of column in prices pandas DataFrame used as parameter to the `fit` method
		"""
	def fit(self, prices, change_points):
		"""
		Finds indexes where the barrier is touched, calculates labels
		
		Parameters
		----------
		prices: Pandas DataFrame
			dataset with aggregated prices
		change_points: array
			array of indices with change points
			
		"""
		pass
	def transform(self):
		"""
		Returns calculated values
		
		Returns
		-------
			Pandas Dataframe with the following columns: 
				index : timestamps same as returned by the change point detection algorithm
				col1 : index of event when the barrier was touched
				col2 : label (upper barrier touched, lower barrier touched, time expired without touching barrier)		
		"""
		pass

	def fit_transfom(self, X):
		"""
		calls the fit and the transofmr methods sequentially
		"""
		pass
	def update(self,prices, change_points):
		"""
		Performs update only for the new values since the last run of fit_transform and concatenates the results to the DataFrame previously returned by the `transform` method.
		"""
		pass
class DatasetBuilder(_ComplexToTabularTransformer):
	"""
	Builds the dataset in the correct format that can be used by the estimator at the end of the pipeline, e.g. tabular, time series, etc.
	
	
	Parameters
	---------
	lookback : int
		Lookback in number of periods counting from each changepoint event
	"""
	def __init__(self, lookback):
		pass
	
	def fit(self,x_input,y_input):
		"""
		x_input: dictionary
			dictionary with values from which the Dataset Builder will be able to compile the features of the dataset
		y_input : dictionary
			dictionary with values from which the Dataset Builder will be able to compile the labels of the dataset
		"""
		pass
	def transform(self):
		"""
		Returns the result of `fit`
		"""
	def fit_transform(self):
		"""
		Calls sequentially fit and transform
		"""
		pass
	def update(self, x_input, y_input):
		"""
		Updates the dataset with the new values since last run
		"""
		pass

class OnlineUnsupervisedPipeline(BaseEstimator):
	"""
	The pipeline will act as a `glue` between the steps.
	
	Main requirements
	1. Check steps for consistency
	2. Create dictionary with  parameters from inputs. Instantiate steps, call functions using synthax step_func(**created_dict), where the keys of created_dict correspond to step_func arguments.
	"""
	def fit(self, X):
		pass
	def predict(self):
		pass
	def fit_predict(X):
		pass
	def update(X):
		pass
```

**Comment by mloning** [comment]
@fkiraly @ViktorKaz I'm happy to integrate this functionality, but we need to break it down into smaller work packages/PRs.

**Comment by mloning** [comment]
> ```python
> pipeline = OnlineUnsupervisedPipeline(
>     [
>         ('feature_extraction', DollarBars(price_col ='Price', volume_col ='Volume'), input_values={'X':'original_series'} ), # input_values={'X':'original_series'} can be omitted
>         ('changepoint_detection',CUSUM(), input_values={'X':'feature_extraction'}),
>         ('labeling', TripleBarrier(price_col='closing_price'), input_values={ 'change_points':'changepoint_detection', 'prices':'feature_extraction'}), 
>         ('build_dataset', DatasetBuilder(lookback=20), input_values={ 'untransformed_series':'original_series','changepoint_detection':'changepoint_detection'}), 
>         ('estimator',SVC(), input_values={'X':'build_dataset','y':'labelling'})
>     ]
> )
> pipeline.fit(X)
> ```

For example, the above code for example seems to combine annotation and reduction from (directional) forecasting to tabular classification.  We need to separately define APIs for annotation and directional forecasting including the reduction to classification.

**Comment by ViktorKaz** [comment]
Do you want to start by integrating the [pipeline object](https://github.com/ViktorKaz/sktime/blob/pipeline/sktime/pipeline/pipeline.py)? I can create a PR just for it.

**Comment by fkiraly** [comment]
Yes, agreed with @mloning - this touches estimators of multiple scientific types (directional forecasters, change point detectors), for which we may like to introduce support and template separately. Would it be a good idea to start with directional forecasters, before adding a pipeline for directional forecasters?

**Comment by mloning** [comment]
> Do you want to start by integrating the [pipeline object](https://github.com/ViktorKaz/sktime/blob/pipeline/sktime/pipeline/pipeline.py)? I can create a PR just for it.

The pipeline seems to be the most complex object, combining multiple APIs (reduction and directional forecasting). So I would start with simple components, either directional forecasters or annotators. 

Since there is some ongoing work on annotation, directional forecasting seems like a good starting point. 

I suggest to implement the following first: 
* an abstract base class: `BaseDirectionalForecaster` which defines the interface (should be the same as `BaseForecaster` except that predict returns a series of categorical values if I understand this correctly)
* a concrete algorithm (if you want to start with reduction to classification, see `make_reduction(...)` in `sktime/forecasting/compose/_reduce.py`)
* a loss function to evaluate categorical forecasts (see #718)
* a unit testing framework, see e.g. `sktime/forecasting/tests/test_all_forecasters.py`
* a short example notebook for documentation
